include/vcpkg/base/util.h

```diff
@@ -132,22 +132,40 @@ namespace vcpkg::Util
         return it->second;
     }
 
-    template<class Map, class Key>
-    const typename Map::mapped_type& value_or_default(const Map& map,
-                                                      Key&& key,
-                                                      const typename Map::mapped_type& default_value)
+    inline void assign_if_set_and_nonempty(std::string& target,
+                                           const std::unordered_map<std::string, std::string>& haystack,
+                                           StringLiteral needle)
     {
-        const auto it = map.find(static_cast<Key&&>(key));
-        if (it == map.end())
+        auto it = haystack.find(needle.to_string());
+        if (it != haystack.end() && !it->second.empty())
         {
-            return default_value;
+            target = it->second;
         }
+    }
 
-        return it->second;
+    template<class T>
+    void assign_if_set_and_nonempty(Optional<T>& target,
+                                    const std::unordered_map<std::string, std::string>& haystack,
+                                    StringLiteral needle)
+    {
+        auto it = haystack.find(needle.to_string());
+        if (it != haystack.end() && !it->second.empty())
+        {
+            target.emplace(it->second);
+        }
     }
 
-    template<class Map, class Key>
-    void value_or_default(const Map& map, Key&& key, typename Map::mapped_type&& default_value) = delete;
+    inline const std::string* value_if_set_and_nonempty(const std::unordered_map<std::string, std::string>& haystack,
```

Comment:
Why not use optional here?

Reply:
Optional would imply copying the content, and we don't need to copy the content. The string is already in the unordered_map, so we can just use a pointer to it.

Optional is only necessary in cases where the thing must own the actual data.